numba/function.py

```diff
@@ -0,0 +1,204 @@
+from llvmlite import ir
+from numba.types import (
+    FunctionType, FunctionProtoType, WrapperAddressProtocol)
+
+
+@typeof_impl.register(WrapperAddressProtocol)
+def typeof_WrapperAddressProtocol(val, c):
+    return val.signature().as_type()
+
+
+@typeof_impl.register(CFunc)
+def typeof_CFunc(val, c):
+    return val._sig.as_type()
+
+
+# TODO: typeof_impl for Dispatcher, types.FunctionType, ctypes.CFUNCTYPE
+
+
+@register_model(FunctionProtoType)
+class FunctionProtoModel(models.PrimitiveModel):
+    """FunctionProtoModel describes the signatures of first-class functions
+    """
+    def __init__(self, dmm, fe_type):
+        if isinstance(fe_type, FunctionType):
+            ftype = fe_type.ftype
+        elif isinstance(fe_type, FunctionProtoType):
+            ftype = fe_type
+        else:
+            raise NotImplementedError((type(fe_type)))
+        retty = dmm.lookup(ftype.rtype).get_value_type()
+        args = [dmm.lookup(t).get_value_type() for t in ftype.atypes]
+        be_type = ir.PointerType(ir.FunctionType(retty, args))
+        super(FunctionProtoModel, self).__init__(dmm, fe_type, be_type)
+
+
+@register_model(FunctionType)
+class FunctionModel(models.StructModel):
+    """FunctionModel holds addresses of function implementations
+    """
+    def __init__(self, dmm, fe_type):
+        members = [
+            # address of cfunc-ded function:
+            ('addr', nbtypes.voidptr),
+            # address of PyObject* referencing the Python function
+            # object, currently it is CFunc:
+            ('pyaddr', nbtypes.voidptr),
+        ]
+        super(FunctionModel, self).__init__(dmm, fe_type, members)
+
+
+@lower_constant(FunctionType)
+def lower_constant_function_type(context, builder, typ, pyval):
+    if isinstance(pyval, CFunc):
+        addr = pyval._wrapper_address
+        sfunc = cgutils.create_struct_proxy(typ)(context, builder)
+        llty = context.get_value_type(nbtypes.voidptr)
+        sfunc.addr = builder.inttoptr(ir.Constant(ir.IntType(64), addr), llty)
+        # TODO: is incref(pyval) needed? See also related comments in
+        # unboxing below.
+        sfunc.pyaddr = builder.inttoptr(
+            ir.Constant(ir.IntType(64), id(pyval)), llty)
+        return sfunc._getvalue()
+
+    # TODO: implement support for Dispatcher, WrapperAddressProtocol,
+    # and types.FunctionType, ctypes.CFUNCTYPE
+
+    raise NotImplementedError(
+        'lower_constant_struct_function_type({}, {}, {}, {})'
+        .format(context, builder, typ, pyval))
+
+
+def _get_wrapper_address(func, sig):
+    """Return the address of a compiled function that implements `func`
+    algoritm.
+
+    Warning: The compiled function must be compatible with the given
+    signature `sig`. If it is not, then calling the compiled function
+    will likely crash the program with segfault.
+
+    Parameters
+    ----------
+    func : object
+      Specify a function object that can be numba.cfunc decorated or
```

Comment:
The `func` can also be Python function. When this is the case, the mapping `func = cfunc(sig)(func)` will be applied.

`func` must be compatible with `sig`, in general, see the Warning note. There are no compatibility checks in order to keep `_get_wrapper_address` light-weight. In fact, compatibility is ensured at the point of calling the njitted function with function arguments so that `_get_wrapper_address` is good (in most cases).

Reply:
Added more notes to documentation string.